### PR TITLE
Disable tile+distribute for elementwise operations due to regressions on CPU

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
To use tile+distribute the tile size configuration needs to be selected so as to not have significant perf regressions.

### CODE REVIEW

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp
Comment (antiagainst) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:952]: Sets elementwise operations ...

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp
Comment (antiagainst) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:955]: Is it really necessary to template this? FWICT, you can just check `isa<GenericOp, IndexedGenericOp>()`, right?

**Review by antiagainst — APPROVED** [review]
Checked dispatch region formation on MobileNet: with this we are creating 94 dispatches, 1 more than previous. Looks two consecutive `linalg.generic` is not fused with this change. But it almost has no impact on performance so I won't block this for it. 

Still I have concerns over this. From what I can tell, this is a workaround, with the cost of complicating the logic in dispatch region formation quite a bit. Do we have an idea what the proper fix is? How far away are we from the real fix? Is it possible to just land the proper fix instead of having this entirely?

**Review by asaadaldien — APPROVED** [review]

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp
Comment (asaadaldien) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:53]: Ah this is independent from enabling tile-and-distribute for element wise ops but was triggered by disabling it ?

**Comment by MaheshRavishankar** [comment]
> Checked dispatch region formation on MobileNet: with this we are creating 94 dispatches, 1 more than previous. Looks two consecutive `linalg.generic` is not fused with this change. But it almost has no impact on performance so I won't block this for it.

If you could get me the two ops that arent fused that would really help.

> 
> Still I have concerns over this. From what I can tell, this is a workaround, with the cost of complicating the logic in dispatch region formation quite a bit. Do we have an idea what the proper fix is? How far away are we from the real fix? Is it possible to just land the proper fix instead of having this entirely?

My take is the opposite here. THe previous dispatch region creation logic (of how fusion is done) was very complicated and didnt really account for the fact that some of the features (like fusion of named ops with elementwise ops and tile + distribute of elementwise ops) arent plumbed through fully and have some rough edges. These need proper fixes downstream
- Making sure that fusion of named ops with elementwise ops works in all cases by using vectorization and load-store forwarding
- Making sure that the vectorization is done for all elementwise operation so that the tile+distribute doesnt cause regressions cause the llvm auto-vectorizer doesnt kick in on the generated code.
Given that this needs to be addressed properly it make sense to separate the core logic of dispatch region creation with the logic needed to enable these features so that they can be separated concerns.

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp
Comment (MaheshRavishankar) [reply iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:53]: Yeah, just adding comments here that these options should be removed soon.

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp
Comment (MaheshRavishankar) [reply iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:955]: unfortunately yes. It seems like the tile+distribute was done only for generic operations and not indexed generic operations. So I am keeping the same functionality here for the GPU side of things.
So when tile+distribute of elementwise operations is enabled its done only for generic ops. But after that to use the fallback dispatch region creation you need to make both indexed generic and generic ops root ops, so that fill ops can be fused with them.

**Comment by antiagainst** [comment]
> > Checked dispatch region formation on MobileNet: with this we are creating 94 dispatches, 1 more than previous. Looks two consecutive `linalg.generic` is not fused with this change. But it almost has no impact on performance so I won't block this for it.
> 
> If you could get me the two ops that arent fused that would really help.

Sure! Here it is: https://gist.github.com/antiagainst/c0161fb817550ff19fb89427cf31775c#file-mobile-fusion-mlir-L9821

At the current top of the tree, `%88` is fused with `%89`, which just performs a division over a constant.

> 
> > Still I have concerns over this. From what I can tell, this is a workaround, with the cost of complicating the logic in dispatch region formation quite a bit. Do we have an idea what the proper fix is? How far away are we from the real fix? Is it possible to just land the proper fix instead of having this entirely?
> 
> My take is the opposite here. THe previous dispatch region creation logic (of how fusion is done) was very complicated and didnt really account for the fact that some of the features (like fusion of named ops with elementwise ops and tile + distribute of elementwise ops) arent plumbed through fully and have some rough edges. These need proper fixes downstream
> 
> * Making sure that fusion of named ops with elementwise ops works in all cases by using vectorization and load-store forwarding
> * Making sure that the vectorization is done for all elementwise operation so that the tile+distribute doesnt cause regressions cause the llvm auto-vectorizer doesnt kick in on the generated code.
>   Given that this needs to be addressed properly it make sense to separate the core logic of dispatch region creation with the logic needed to enable these features so that they can be separated concerns.

Thanks for the detailed explanation! This is exactly what I wanted to know: the difficulties that require this kind of separation. From the list it's more clear to me that it will take quite some time to handle them. Although I don't think the existing dispatch formation is complicated compared to the multiple steps we have here and IIUC it's still what we want to get back to eventually, after landing this patch and addressing the issues above.

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp
Comment (antiagainst) [reply iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:956]: Ah, right. I recall now. Thanks.

**Comment by MaheshRavishankar** [comment]
> At the current top of the tree, `%88` is fused with `%89`, which just performs a division over

Strange. That should not be fused. The producer has a "reduction" iterator type and the consumer is parallel. Those should actually not be fused at all (even before this PR) unless I am missing something

**Comment by MaheshRavishankar** [comment]
> Thanks for the detailed explanation! This is exactly what I wanted to know: the difficulties that require this kind of separation. From the list it's more clear to me that it will take quite some time to handle them. Although I don't think the existing dispatch formation is complicated compared to the multiple steps we have here and IIUC it's still what we want to get back to eventually, after landing this patch and addressing the issues above.

It is complicated cause its all inter-twined. When everything works as expected I agree what was there before would be what we would do. But now some of the decision that are related to fusion and tile + distribute of elementwise operations are all inter-twined, and leaks into the "default" case of not doing these as well.
I wouldnt have done this refactoring if it was separable enough to avoid doing this, but I kept hitting issues where the decision of one leaked into the other. So I mainly wanted to make them truly disjoint.

**Comment by antiagainst** [comment]
> > Thanks for the detailed explanation! This is exactly what I wanted to know: the difficulties that require this kind of separation. From the list it's more clear to me that it will take quite some time to handle them. Although I don't think the existing dispatch formation is complicated compared to the multiple steps we have here and IIUC it's still what we want to get back to eventually, after landing this patch and addressing the issues above.
> 
> It is complicated cause its all inter-twined. When everything works as expected I agree what was there before would be what we would do. But now some of the decision that are related to fusion and tile + distribute of elementwise operations are all inter-twined, and leaks into the "default" case of not doing these as well.
> I wouldnt have done this refactoring if it was separable enough to avoid doing this, but I kept hitting issues where the decision of one leaked into the other. So I mainly wanted to make them truly disjoint.

Understood and agreed.

**Comment by antiagainst** [comment]
> > At the current top of the tree, `%88` is fused with `%89`, which just performs a division over
> 
> Strange. That should not be fused. The producer has a "reduction" iterator type and the consumer is parallel. Those should actually not be fused at all (even before this PR) unless I am missing something

Interesting. I think it should be fine to fuse these two? Is there a correctness issue you see here? The reduction is scoped inside the tiled loops and we distribute along the parallel dimension.

**Comment by MaheshRavishankar** [comment]
you mean tile and fuse this? This falls into the same case of matmul/conv + elementwise fusion as well. So if we want to enable this, then we need to treat generic op with reduction iterator types as root ops too. But again those will not get vectorized. So on balance lets leave it this way.

There is another solution where we can fuse then during elementwise op fusion. We can do this if the indexing map in the consumer is an identity map. The fused op will become an indexed generic operation. That might be a better approach.

**Pull request merged**
